### ISSUE TITLE
Can this be run like a server similar to cellxgene-gateway?

### ISSUE THREAD

**Issue by eliko1991** [bag-cnag/cxg_on_k8#1]
Hi,

Not sure if I am just too noobish with both docker and cellxgene, but I am trying to run the VIP slim version like a server similar to cellxgene-gateway.

I tried building and running similar to https://github.com/Novartis/cellxgene-gateway?tab=readme-ov-file#running-cellxgene-gateway-with-docker but unsuccesful.

building vip slim docker file
`docker build -t cellxgene_vip . `
running the newly build image
`docker run --privileged -it -d --rm -v /data:/cellxgene-data -p 5005:5005 cellxgene_vip`

I am trying to run this with docker on my server, appreciate any help.

And thank you for creating these dockerfiles, they require a lot of effort and time!

Eli

**Comment by Neah-Ko** [comment]
Hello @eliko1991 ,

### Cellxgene
- Is an app running a web server allowing you to visualize `one` file in `.h5ad` format.
- Cellxgene VIP is a patched version from this team: https://github.com/interactivereport/cellxgene_VIP that adds a bunch of functionalities (plots, notebooks, ...).
  - Just like cellxgene, it is meant to visualize one file at a time.


To build use:
`docker build . -f docker/Dockerfile_cellxgene_VIP_slim -t cellxgene:1.1.2-VIP-python3.8-slim-bookworm`
To run use:
`docker run -it -p 5005:5005 -v /path/to/data:/data cellxgene:1.1.2-VIP-python3.8-slim-bookworm /data/myfile.h5ad`
Then you should be able to access through a web browser at `yourserver_ip/:5005`



### cellxgene gateway
> Cellxgene Gateway allows you to use the Cellxgene Server provided by the Chan Zuckerberg Institute (https://github.com/chanzuckerberg/cellxgene) with multiple datasets.

I haven't analyzed it in much detail, but it seems like a different product running cellxgene instances in subprocesses to serve multiple datasets.  If you want to use it in combination with cellxgene VIP you may try to to patch it. That may or may not be simple.


### What this repo is about ?
This repo is a proof of concept, for a similar idea. However, leveraging a kubernetes cluster instead to run cellxgene instances on demand.
Ultimately, there'll be a frontend listing datasets and the possibility to visualize them individually.




Let me know if you have any more doubts.
In case not, you may send me the outputs of the commands you're using.

Best,

**Comment by eliko1991** [comment]
Hi @Neah-Ko ,


Thanks for the very detailed response, clears a number of things up for me, extremely appreciated.

My main experience with cellxgene was through server, but I can see how launching per h5ad is powerfull as well. Haven't considered the ondemand approach but I like it a lot, will try to explore it.

I attempted to run the image per your suggestion, had to add `launch` before the h5ad path, the container seems to run fine but the application doesn't load when navigating to it. I tried both on my server and my local laptop, and I can't reach it via server-ip:5005 or localhost:5005. Any tips appreciated.

Thank you again for taking the time to respond to a stranger on the internet!


Eli

**Comment by Neah-Ko** [comment]
Hello, @eliko1991 

Could you provide me with:
- Cellxgene logs
When running from the command line you should have an output looking like that:
```bash
$ docker run -p 5005:5005 -it cellxgene:1.1.2-VIP-python3.8-slim-bookworm launch --verbose --host 0.0.0.0 https://github.com/chanzuckerberg/cellxgene/raw/main/example-dataset/pbmc3k.h5ad 
There's a new version of cellxgene available (1.2.0)!
To upgrade, run the following: pip install --upgrade cellxgene

[cellxgene] Starting the CLI...
[cellxgene] Loading data from pbmc3k.h5ad.
[cellxgene] Warning: Moving element from .uns['neighbors']['distances'] to .obsp['distances'].

This is where adjacency matrices should go now. 
[cellxgene] Warning: Moving element from .uns['neighbors']['connectivities'] to .obsp['connectivities'].

This is where adjacency matrices should go now. 
[cellxgene] Launching! Please go to http://0.0.0.0:5005 in your browser.
[cellxgene] Type CTRL-C at any time to exit.
 * Serving Flask app 'server.app.app'
 * Debug mode: off
INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:5005
 * Running on http://172.17.0.3:5005
INFO:werkzeug:Press CTRL+C to quit
```
-> Mind the `--host 0.0.0.0` parameter that tends to be better accepted than `localhost` by various networking tools.
Hit the webpage at `http://0.0.0.0:5005`

Then visiting the webpage, shows API endpoints hits.

- Console logs
On your browser, hit f12 and let me know if you see anything in the network tab.

Best,
